pkg/monitor/monitor.go

```diff
@@ -80,6 +87,61 @@ func (m *Monitor) isSameNode(k8sNodeName string, driverNode *volume.NodeInfo) bo
 	return volume.IsNodeMatch(node, driverNode)
 }
 
+func (m *Monitor) podMonitor() error {
+	fn := func(object runtime.Object) error {
+		pod, ok := object.(*v1.Pod)
+		if !ok {
+			err := fmt.Errorf("invalid object type on pod watch: %v", object)
+			return err
+		}
+
+		podUnknownState := false
+
+		if pod.Status.Reason == node.NodeUnreachablePodReason {
+			podUnknownState = true
+		} else if pod.ObjectMeta.DeletionTimestamp != nil {
+			n, err := k8s.Instance().GetNodeByName(pod.Spec.NodeName)
+			if err != nil {
+				return err
+			}
+
+			// Check if node has eviction taint
+			for _, taint := range n.Spec.Taints {
+				if taint.Key == algorithm.TaintNodeUnreachable &&
+					taint.Effect == v1.TaintEffectNoExecute {
+					podUnknownState = true
+					break
+				}
+			}
+		}
+
+		if podUnknownState {
+			owns, err := m.doesDriverOwnPodVolumes(pod)
+			if err != nil || !owns {
+				return nil
+			}
+
+			storklog.PodLog(pod).Infof("Force deleting pod as it's in unknown state.")
+
+			// force delete the pod
+			err = k8s.Instance().DeletePods([]v1.Pod{*pod}, true)
+			if err != nil {
+				storklog.PodLog(pod).Errorf("Error deleting pod: %v", err)
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	if err := k8s.Instance().WatchPods("", fn); err != nil {
```

Comment:
No, this will handle the later case too. Just tested this.

1. Stop stork
2. Start statefulset pod
3. Stop kubelet and wait for pod to get into terminating evicting state
4. Start stork

After 4, stork found the pod in unknown state and deleted it.

Logs below.

```
kubectl get pods -n kube-system -l name=stork
No resources found.
```

```
kubectl get pods -owide
NAME           READY   STATUS    RESTARTS   AGE    IP           NODE   NOMINATED NODE   READINESS GATES
mysql-sset-0   1/1     Running   0          104s   10.36.0.10   k3n1   <none>           <none>
```

```
kubectl get node k3n1 -o yaml | grep -A 5 taint
  taints:
  - effect: NoSchedule
    key: node.kubernetes.io/unreachable
    timeAdded: "2019-03-27T22:42:24Z"
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
```

```

```
kubectl get pods -owide
NAME           READY   STATUS        RESTARTS   AGE     IP           NODE   NOMINATED NODE   READINESS GATES
mysql-sset-0   1/1     Terminating   0          9m54s   10.36.0.10   k3n1   <none>           <none>
```

Now start stork

```
time="2019-03-27T22:51:46Z" level=debug msg="Monitoring storage nodes"
time="2019-03-27T22:51:46Z" level=info msg="Force deleting pod as it's in unknown state." Namespace=default Owner=StatefulSet/mysql-sset PodName=mysql-sset-0
time="2019-03-27T22:51:46Z" level=debug msg="Pod doesn't have any volumes by driver" Namespace=kube-system Owner=ReplicaSet/coredns-86c58d9df4 PodName=coredns-86c58d9df4-qfnz4
time="2019-03-27T22:51:46Z" level=info msg="Force deleting pod as it's in unknown state." Namespace=default Owner=StatefulSet/mysql-sset PodName=mysql-sset-0
time="2019-03-27T22:51:46Z" level=info msg="Force deleting pod as it's in unknown state." Namespace=default Owner=StatefulSet/mysql-sset PodName=mysql-sset-0
time="2019-03-27T22:51:51Z" level=info msg="Registering CRDs"
I0327 22:51:51.197637       1 snapshot-controller.go:167] Starting snapshot controller
```